### PR TITLE
Only quote non-syntactic and reserved words

### LINKED ISSUE #3704: Make `dbQuoteIdentifier()` only quote when necessary?
For teaching, it'd be cool if instead of:

```sql
SELECT "carat", "cut", "color", "clarity", "price"
FROM "diamonds"
WHERE ("price" > 15000.0)
```

duckdb could generate:

```SQL
SELECT carat, cut, color, clarity, price
FROM diamonds
WHERE ("price" > 15000.0)
```

I think duckdb is better placed to do this than the other database backends since it's easier for you to keep the reserved words in sync. 

I'd be happy to implement this if you can give a few hints as to how to reliably form a list of the reserved words and otherwise what constraints a table/field name must follow. (I'm guessing only letters + digits + underscore, and must start with a letter. But is letter ASCII or unicode?) (Do you allow (eg.) U+FE4D, the dashed low-line, which I think the standard allows?)

### PR DESCRIPTION
**Pull request by hadley**
Fixes #3704

Let me know if I missed any project conventions.

### CODE REVIEW

File: tools/rpkg/R/dbQuoteIdentifier__duckdb_connection.R [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+#' @rdname duckdb_connection-class
+#' @export
+setMethod("dbQuoteIdentifier", signature("duckdb_connection"), dbQuoteIdentifier__duckdb_connection)
+
+setMethod("dbQuoteIdentifier", signature("duckdb_connection", "character"), dbQuoteIdentifier__duckdb_connection)
+setMethod("dbQuoteIdentifier", signature("duckdb_connection", "SQL"), dbQuoteIdentifier__duckdb_connection)
+setMethod("dbQuoteIdentifier", signature("duckdb_connection", "Id"), dbQuoteIdentifier__duckdb_connection)
+
+reserved_words <- function(con) {
```
Comment (hadley) [inline tools/rpkg/R/dbQuoteIdentifier__duckdb_connection.R:23]: This caches the list of reserved words on first run; let me know if you'd prefer a different strategy.

**Comment by hadley** [comment]
Hmmm, I can't run the tests myself because I get:

```
*** caught illegal operation ***
address 0x11a262598, cause 'illegal opcode'

Traceback:
 1: Array__GetScalar(Array$create(x, type = type), 0)
 2: Scalar$create(x)
 3: compute___expr__scalar(Scalar$create(x))
 4: Expression$scalar(TRUE)
 5: self$Scan(.data)
 6: plan$Build(.data)
 7: do_exec_plan(x)
 8: doTryCatch(return(expr), name, parentenv, handler)
 9: tryCatchOne(expr, names, parentenv, handlers[[1L]])
10: tryCatchList(expr, classes, parentenv, handlers)
11: tryCatch(tab <- do_exec_plan(x), error = function(e, call = caller_env(n = 4)) {    handle_csv_read_error(e, x$.data$schema, call)})
12: collect.arrow_dplyr_query(.)
13: collect(.)
```

Any ideas what's causing this?

File: tools/rpkg/R/dbQuoteIdentifier__duckdb_connection.R (hunk 1 above)
Comment (hadley) [reply tools/rpkg/R/dbQuoteIdentifier__duckdb_connection.R:23]: Oh hmmm, of course, this doesn't work with the `translate_sql()` tests which pass in a dummy connection. Maybe it's better to do this `.onLoad`?

**Review by krlmlr — APPROVED** [review]
Looks good!

Should we do a memoise-like function instead of eagerly querying the reserved words?

```r
make_get_reserved_words <- function() {
  reserved_words <- NULL
  function() {
    if (is.null(reserved_words)) {
      reserved_words <<- ...
    }
    reserved_words
  }
}

get_reserved_words <- make_get_reserved_words()
```

We could also store in `the` if you feel strongly about it.

**Comment by hadley** [comment]
Oh yeah, that's a better way of resolving the problems I was having.

**Comment by Mytherin** [comment]
It’s not immediately clear to me either, @krlmlr could you have a look?

**Comment by krlmlr** [comment]
Now running a reduced version at https://github.com/krlmlr/duckdb/actions/runs/2465657322 to check if I can replicate the covr failures.

**Comment by krlmlr** [comment]
The covr package doesn't particularly like the dynamic querying of the keywords. It works with a static list, e.g. https://github.com/krlmlr/duckdb/commit/1b1655e3ad17a07135c05bac5ce2fb044cf7993b scrapes `kwlist.hpp` and generates the list of keywords as part of `./configure` .

**Comment by krlmlr** [comment]
FWIW, the generated `R/reserved.R` file looks like this:

```r
# Generated by ./configure, do not edit by hand
reserved_words <- function() {
c("abort", "absolute", "access", "action", "add", "admin", "after", 
"aggregate", "all", "also", "alter", "always", "analyse", "analyze", 
"and", "any", "array", "as", "asc", "assertion", "assignment", 
"asymmetric", "at", "attach", "attribute", "authorization", "backward", 
"before", "begin", "between", "bigint", "binary", "bit", "boolean", 
"both", "by", "cache", "call", "called", "cascade", "cascaded", 
"case", "cast", "catalog", "chain", "char", "character", "characteristics", 
"check", "checkpoint", "class", "close", "cluster", "coalesce", 
"collate", "collation", "column", "columns", "comment", "comments", 
"commit", "committed", "compression", "concurrently", "configuration", 
"conflict", "connection", "constraint", "constraints", "content", 
"continue", "conversion", "copy", "cost", "create", "cross", 
"csv", "cube", "current", "current_catalog", "current_date", 
"current_role", "current_schema", "current_time", "current_timestamp", 
"current_user", "cursor", "cycle", "data", "database", "day", 
"days", "deallocate", "dec", "decimal", "declare", "default", 
"defaults", "deferrable", "deferred", "definer", "delete", "delimiter", 
"delimiters", "depends", "desc", "describe", "detach", "dictionary", 
"disable", "discard", "distinct", "do", "document", "domain", 
"double", "drop", "each", "else", "enable", "encoding", "encrypted", 
"end", "enum", "escape", "event", "except", "exclude", "excluding", 
"exclusive", "execute", "exists", "explain", "export", "export_state", 
"extension", "external", "extract", "false", "family", "fetch", 
"filter", "first", "float", "following", "for", "force", "foreign", 
"forward", "freeze", "from", "full", "function", "functions", 
"generated", "glob", "global", "grant", "granted", "group", "grouping", 
"grouping_id", "handler", "having", "header", "hold", "hour", 
"hours", "identity", "if", "ignore", "ilike", "immediate", "immutable", 
"implicit", "import", "in", "including", "increment", "index", 
"indexes", "inherit", "inherits", "initially", "inline", "inner", 
"inout", "input", "insensitive", "insert", "install", "instead", 
"int", "integer", "intersect", "interval", "into", "invoker", 
"is", "isnull", "isolation", "join", "json", "key", "label", 
"language", "large", "last", "lateral", "leading", "leakproof", 
"left", "level", "like", "limit", "listen", "load", "local", 
"localtime", "localtimestamp", "location", "lock", "locked", 
"logged", "macro", "map", "mapping", "match", "materialized", 
"maxvalue", "method", "microsecond", "microseconds", "millisecond", 
"milliseconds", "minute", "minutes", "minvalue", "mode", "month", 
"months", "move", "name", "names", "national", "natural", "nchar", 
"new", "next", "no", "none", "not", "nothing", "notify", "notnull", 
"nowait", "null", "nullif", "nulls", "numeric", "object", "of", 
"off", "offset", "oids", "old", "on", "only", "operator", "option", 
"options", "or", "order", "ordinality", "out", "outer", "over", 
"overlaps", "overlay", "overriding", "owned", "owner", "parallel", 
"parser", "partial", "partition", "passing", "password", "percent", 
"placing", "plans", "policy", "position", "pragma", "preceding", 
"precision", "prepare", "prepared", "preserve", "primary", "prior", 
"privileges", "procedural", "procedure", "program", "publication", 
"qualify", "quote", "range", "read", "real", "reassign", "recheck", 
"recursive", "ref", "references", "referencing", "refresh", "reindex", 
"relative", "release", "rename", "repeatable", "replace", "replica", 
"reset", "respect", "restart", "restrict", "returning", "returns", 
"revoke", "right", "role", "rollback", "rollup", "row", "rows", 
"rule", "sample", "savepoint", "schema", "schemas", "scroll", 
"search", "second", "seconds", "security", "select", "sequence", 
"sequences", "serializable", "server", "session", "session_user", 
"set", "setof", "sets", "share", "show", "similar", "simple", 
"skip", "smallint", "snapshot", "some", "sql", "stable", "standalone", 
"start", "statement", "statistics", "stdin", "stdout", "storage", 
"stored", "strict", "strip", "struct", "subscription", "substring", 
"summarize", "symmetric", "sysid", "system", "table", "tables", 
"tablesample", "tablespace", "temp", "template", "temporary", 
"text", "then", "time", "timestamp", "to", "trailing", "transaction", 
"transform", "treat", "trigger", "trim", "true", "truncate", 
"trusted", "try_cast", "type", "types", "unbounded", "uncommitted", 
"unencrypted", "union", "unique", "unknown", "unlisten", "unlogged", 
"until", "update", "user", "using", "vacuum", "valid", "validate", 
"validator", "value", "values", "varchar", "variadic", "varying", 
"verbose", "version", "view", "views", "virtual", "volatile", 
"when", "where", "whitespace", "window", "with", "within", "without", 
"work", "wrapper", "write", "xml", "xmlattributes", "xmlconcat", 
"xmlelement", "xmlexists", "xmlforest", "xmlnamespaces", "xmlparse", 
"xmlpi", "xmlroot", "xmlserialize", "xmltable", "year", "years", 
"yes", "zone")
}
```

Not pretty, but serves the job.

**Comment by hannes** [comment]
Would it not be easiest to query the duckdb_keywords() table function? e.g. 
```
D select * from duckdb_keywords() limit 10;
┌──────────────┬──────────────────┐
│ keyword_name │ keyword_category │
├──────────────┼──────────────────┤
│ abort        │ unreserved       │
│ absolute     │ unreserved       │
│ access       │ unreserved       │
│ action       │ unreserved       │
│ add          │ unreserved       │
│ admin        │ unreserved       │
│ after        │ unreserved       │
│ aggregate    │ unreserved       │
│ all          │ reserved         │
│ also         │ unreserved       │
└──────────────┴──────────────────┘
```

**Comment by krlmlr** [comment]
That's what this PR does during run time, it fails the coverage check for reasons that are hard to understand.

Querying during build or install time is hard because we don't necessarily have duckdb ready to use by then? What am I missing?

What is it that you don't seem to like about the source code transformation?

**Comment by hadley** [comment]
I can reproduce the coverage failure locally:

```
Backtrace:
    ▆
 1. └─devtools:::test_coverage()
 2.   └─covr::package_coverage(pkg$path, ...)
 3.     └─covr:::merge_coverage(trace_files)
 4.       ├─base::suppressWarnings(readRDS(files[1]))
 5.       │ └─base::withCallingHandlers(...)
 6.       └─base::readRDS(files[1])
 ```

**Comment by krlmlr** [comment]
I'll take a closer look at why covr is failing.

**Comment by Mytherin** [comment]
@krlmlr any updates here?

**Comment by krlmlr** [comment]
On it now, sorry for the delay.

**Comment by krlmlr** [comment]
`devtools:::test_coverage(type = "none", clean = FALSE)` works. Will bisect from here to see try to understand the trigger.

**Comment by krlmlr** [comment]
Found the cause: an ALTREP vector that is already freed when covr tries to collect the data.

> This looks like it has been freed.

We can materialize with `rev(rev(.))` or similar, this avoids the errors.

Alternatively (or in addition), we could make the reserved words a component/slot in the connection object, they might be different from connection to connection in theory?

Sending an updated PR tomorrow.

**Pull request closed without merging**